### PR TITLE
Store: Update product images to use ImageSelector component

### LINKED ISSUE #21302: Store: Implement Media Modal
Right now, Store on WordPress.com uses a simple upload component to upload images. Part of the original designs intended to use the media modal, but we left it out of the initial implementation.

I dug down the rabbit hole again to see how difficult it would be to implement the modal now. It looks like we have some "prior art" now with site icon management (https://wordpress.com/settings/general/:site), which I don't believe existed before.

The site icon implements [post-editor/media-modal](https://github.com/Automattic/wp-calypso/tree/5f15f3365213253724ca0d8798c6cb8160f00bc9/client/post-editor/media-modal) which works OK for selecting a single image, but if you try to allow multiple item selection, the modal takes you through gallery settings and some other screens which are not really applicable outside of the post context.

I think we could role our own `media modal` component that implements `MediaLibrary`, and keeps us out of some of the views we don't want. We can also make it return selected image data in a form that's more useful for passing to WooCommerce API endpoints.

Related: #14760.

### PR DESCRIPTION
**Pull request by joshuatf**
This PR removes the product image uploader and adds in the ImageSelector component.  This allows image selection via the MediaModal and adds a drop zone for images.

This PR depends on #26297 and resolves #21302

### Product Images

1. Select a new image in the MediaModal.  The MediaModal should open when the upload picker is clicked or an existing product image is clicked.
2. Clicking "Set Images" in the MediaModal should correctly show a preview of the selected images.
3. Dragging and dropping an image onto the ImageSelector should upload and set the image.
4. Saving the product by clicking "Update" should persist the data on refreshing the page.  Persisted images should be shown correctly in the ImageSelector preview.
5.  Clicking the "x" (Remove button) in the upper right-hand corner of an image should remove an image.

screen shot 2018-08-15 at 5 38 06 pm

### Product variations

1.  The upload picker for variations should be compact and fit within the space.
2.  Dropping a file on the variation should trigger the drop zone upload.
3.  Clicking on a variation upload picker or existing image should trigger the MediaModal.
4.  Selected images from the MediaModal should appear in the preview area.
5.  Only one image should selectable at a time in the MediaModal.
6. Saving the product by clicking "Update" should persist the variation image on refresh.
7.  Clicking the remove button should remove the variation image.

screen shot 2018-08-15 at 5 38 58 pm

### CODE REVIEW

File: client/blocks/image-selector/preview.jsx [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	};
+
+	renderUploadPlaceholder() {
+		const { onImageClick, translate, compact, multiple } = this.props;
+
+		const classes = classNames( 'image-selector__uploader-wrapper', {
+			compact,
+		} );
+
+		const addString = multiple ? translate( 'Add Images' ) : translate( 'Add Image' );
```
Comment (a8ci18n) [inline client/blocks/image-selector/preview.jsx]: Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/products/product-form-images.js [hunk 2]
```diff
@@ -1,204 +0,0 @@
-/** @format */
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/extensions/woocommerce/app/products/product-form-images.js]: 🌟

File: client/extensions/woocommerce/app/products/product-form-details-card.js [hunk 3]
```diff
@@ -77,7 +76,29 @@ export default class ProductFormDetailsCard extends Component {
 		editProduct( siteId, product, { description } );
 	}
 
-	onImageUpload = image => {
+	setImage = media => {
+		if ( ! media || ! media.items.length ) {
+			return;
+		}
+		const { siteId, product, editProduct } = this.props;
+		const images = media.items.map( item => ( { id: item.ID } ) );
```
[the remark this answers is not in the document]
Comment (joshuatf) [reply client/extensions/woocommerce/app/products/product-form-details-card.js]: You're right, we do.  I like the idea of disabling the save until the actual IDs are passed back.  This should resolve UX issues with saving too early and also prevent saving of transient IDs.

It looks like we can disable the save by dispatching a next step.  Does this like an okay way to implement this? https://github.com/Automattic/wp-calypso/pull/26703/commits/85f14e4a26eb85535b6935bbc8f3748fb715afc7

File: client/extensions/woocommerce/app/products/product-form-images.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshuatf) [reply client/extensions/woocommerce/app/products/product-form-images.js]: ✋

File: client/extensions/woocommerce/app/products/product-form-variations-row.js [hunk 4]
```diff
@@ -78,107 +57,47 @@ class ProductFormVariationsRow extends Component {
...
-	onError = () => {
-		this.setState( {
-			placeholder: null,
-			transientId: null,
-			isUploading: false,
-		} );
+	changeImages = newImages => {
+		const { siteId, editProductVariation, product, variation } = this.props;
+		const image = {
+			id: newImages[ 0 ].ID,
```
[the remark this answers is not in the document]
Comment (joshuatf) [reply client/extensions/woocommerce/app/products/product-form-variations-row.js]: It was definitely possible to save transients on large images.  I believe the above method of disabling the save should resolve this.

**Comment by joshuatf** [comment]
@justinshreve Good catch!  It looked like a z-index issue with the transient overlay.  This commit should fix this - https://github.com/Automattic/wp-calypso/pull/26703/commits/6f95c5da3739e22e252c0b3fe38e498c331964ff

@timmyc I was able to run into the transient issues- thanks for the tip to disable save!

I wasn't able to replicate the issue with portrait oriented images.  Was this from images rotated inside the MediaModal or with metadata?  (Both seemed to work okay for me).  Do you have a sample image I could test with?

**Comment by timmyc** [comment]
@joshuatf sorry got a bit busy/distracted the past week with some other items 😄 

Here is the image that is being shown above: https://timmyc-devsite-timmysflies.blog/wp-content/uploads/2018/04/IMG_20171216_155454468_HDR.jpg

If you can't re-create the problem then I think we are okay with moving forward and getting this merged in.

**Comment by joshuatf** [comment]
@timmyc Thanks for that image!  I was able to recreate the issue and it appears that photon image URLs preserve metadata while they are stripped by default (unless `image_strip_meta ` is set to false) from the WordPress hosted media library.  

I've set the preview images to use the photon URL in https://github.com/Automattic/wp-calypso/pull/26703/commits/b63f2640f771cf541614ec754a74414ed15bd40c

I believe https://github.com/Automattic/wp-calypso/issues/13145 is facing this same issue, but if we're eventually replacing the featured image with the ImageSelector component, this will be resolved.

**Pull request closed without merging**